argon2_out_of_process_support.go

```diff
@@ -0,0 +1,990 @@
+			// eventually returning an error to the caller.
+			var req *Argon2OutOfProcessRequest
+			dec := json.NewDecoder(in)
+			dec.DisallowUnknownFields()
+			if err := dec.Decode(&req); err != nil {
+				// Decoding returned an error.
+				if errors.Is(err, io.EOF) || errors.Is(err, io.ErrUnexpectedEOF) {
+					// The parent has closed their end of the io.Reader, so kill
+					// the tomb normally to begin the process of dying.
+					tmb.Kill(nil)
+					break // Break out of the request processing loop
+				}
+
+				// We failed to decode an incoming request for an unknown reason. We try returning an
+				// error response to the parent - note that we can't set the command field here
+				// because we have no idea what it is.
+				rsp := &Argon2OutOfProcessResponse{
+					ErrorType:   Argon2OutOfProcessErrorUnexpectedInput,
+					ErrorString: fmt.Sprintf("cannot decode request: %v", err),
+				}
+				select {
+				case rspChan <- rsp: // Unbuffered channel, but read end is always there unless the tomb is dying.
+				case <-tmb.Dying():
+					// The tomb began dying before the response was sent.
+				}
+
+				// Break out of the request processing loop.
+				break
+			}
+
+			// We have a request!
+
+			switch req.Command {
+			case Argon2OutOfProcessCommandWatchdog:
+				// Special case to handle watchdog requests.
+				wdRsp, err := watchdog(req.WatchdogChallenge)
+				if err != nil {
+					// As is documented for Argon2OutOfProcessWatchdogHandler, we don't
+					// expect the handler to return an error, so begin the shutdown of
+					// the tomb so that this function eventually returns with an error.
+					return fmt.Errorf("cannot handle watchdog request: %w", err)
+				}
+
+				// Generate the response structure to send to the same goroutine that
+				// will eventually encode and send the KDF response back to the parent.
+				rsp := &Argon2OutOfProcessResponse{
+					Command:          Argon2OutOfProcessCommandWatchdog,
+					WatchdogResponse: wdRsp,
+				}
+				select {
+				case rspChan <- rsp: // Unbuffered channel, but read end is always there unless the tomb is dying.
+				case <-tmb.Dying():
+					// The tomb began dying before the response was sent, so
+					// the outer loop won't run again.
+				}
+			default:
+				// Treat everything else as a KDF request. We don't actually check the
+				// command here - RunArgon2OutOfProcessRequest does this already and will
+				// return an error response if the command is invalid.
+
+				// Spin up a new routine to handle the request, as it blocks and is long running,
+				// and we still want to be able to service watchdog requests on this routine whilst
+				// it's running. Block the current routine until we know the new routine has started
+				// so that the watchdog handler will fail if the new routine doesn't begin properly.
+				startupCh := make(chan struct{})
+
+				tmb.Go(func() error {
+					startupCh <- struct{}{} // Unblock the main routine.
```

Comment:
should have the same effect
```suggestion
					close(startupCh)
```

Reply:
I've made this change now.